docs/adrs/0000-caching-dependencies.md

```diff
@@ -0,0 +1,59 @@
+# 0. Caching dependencies
+Date: 2021-05-21
+
+Status: Proposed
+
+# Context
+`actions/setup-node` is the 2nd most popular action in GitHub Actions. A lot of customers use it in conjunction with [actions/cache](https://github.com/actions/cache) to speed up dependencies installation.  
+See more examples on proper usage in [actions/cache documentation](https://github.com/actions/cache/blob/main/examples.md#node---npm).
+
+# Goals & Anti-Goals
+Integration of caching functionality into `actions/setup-node` action will bring the following benefits for action users:
+- Decrease the entry threshold for using the cache for Node.js dependencies and simplify initial configuration
+- Simplify YAML pipelines because no need additional steps to enable caching
+- More users will use cache for Node.js so more customers will have fast builds!
+
+As the first stage, we will add support for NPM dependencies caching. We can consider adding the same functionality for Yarn later.
+
+We don't persue the goal to provide wide customization of caching in scope of `actions/setup-node` action. The purpose of this integration is covering ~90% of basic use-cases. If user needs flexible customization, we should advice them to use `actions/cache` directly.
+
+# Decision
+- Add `cache` input parameter to `actions/setup-node`. For now, input will accept the following values: 
+  - `npm` - enable caching for npm dependencies
+  - `''` - disable caching (default value)
+  - Potentially, we will be able to extend this input to support Yarn
+- Cache feature will be disabled by default to make sure that we don't break existing customers. We will consider enabling cache by default in next major release (`v3`)
```

Comment:
Considering the fact that users can use either `npm` or `yarn`, I am not sure actually how we can enable it by default in future without making `cache` input mandatory (that doesn't sound good from UX side).
@ahdbilal , any concerns about it?

Reply:
fine with me for not making this on by default. I will get other folks opinions on it.

Reply:
This statement seems appropriate 👍 No need to change the current behavior with `v2`.

We can initially try it out and if it works really well and users see a large improvement then we can consider trying to enable this for `v3`.

One option with `v3` is to make the `cache` input mandatory but we set the default value to `npm`. If there is no `package-lock.json` then the action will output a warning instead of failing. This will not break existing workflows, the majority of users would see improvements due to utilizing cache and for `yarn` users they would only see a warning annotation that should be descriptive enough for them to figure out what is going on. With a mandatory input though, we could add something like a `cache: disabled` option for users who really don't want it. In the future we can make that call but for the moment making it optional is perfect just so it can be tested out.

One potential issue I could see happening with caching being mandatory when moving to v3 is that some users might have extra caching happening that causes them to use more than 5GB. That could cause automatic eviction and all sorts of headaches so we would need to document things extensively before. Lots of considerations but in general I think we should be aggressive about trying to enabling caching by default.

Reply:
I would rather keeping cache disabled even in v3  because don't want to push customers choosing anything. In setup-java, we have seen a lot of feedback that customers don't want to choose between adopt and zulu